core/lib/src/fs/server.rs

```diff
@@ -218,28 +218,35 @@ impl Handler for FileServer {
             .map(|path| self.root.join(path));
 
         match path {
-            Some(p) if p.is_dir() => {
-                // Normalize '/a/b/foo' to '/a/b/foo/'.
-                if options.contains(Options::NormalizeDirs) && !req.uri().path().ends_with('/') {
-                    let normal = req.uri().map_path(|p| format!("{}/", p))
-                        .expect("adding a trailing slash to a known good path => valid path")
-                        .into_owned();
-
-                    return Redirect::permanent(normal)
-                        .respond_to(req)
-                        .or_forward((data, Status::InternalServerError));
-                }
-
-                if !options.contains(Options::Index) {
-                    return Outcome::forward(data, Status::NotFound);
+            Some(mut p) => {
+                if p.is_dir() {
+                    // Normalize '/a/b/foo' to '/a/b/foo/'.
+                    if options.contains(Options::NormalizeDirs) && !req.uri().path().ends_with('/') {
+                        let normal = req.uri().map_path(|p| format!("{}/", p))
+                            .expect("adding a trailing slash to a known good path => valid path")
+                            .into_owned();
+        
+                        return Redirect::permanent(normal)
+                            .respond_to(req)
+                            .or_forward((data, Status::InternalServerError));
+                    }
+        
+                    if !options.contains(Options::Index) {
+                        return Outcome::forward(data, Status::NotFound);
+                    }
+        
+                    p = p.join("index.html");        
                 }
 
-                let index = NamedFile::open(p.join("index.html")).await;
-                index.respond_to(req).or_forward((data, Status::NotFound))
-            },
-            Some(p) => {
-                let file = NamedFile::open(p).await;
-                file.respond_to(req).or_forward((data, Status::NotFound))
+                let gzip_accepted = req.headers().get("Accept-Encoding")
+                    .find(|value| value.contains("gzip"))
```

Comment:
I’m not sure whether a substring match here is a good idea. See https://github.com/tower-rs/tower-http/pull/156/files#diff-83cd763343b90a262fdc4334e5289daea2e9cfb9b1cf76785ea7c42b66d3d5a0R85 for proper parsing code.

Reply:
It depends on what you mean by "good idea" :)

From reading the spec for the header, this should cover what is needed for this change (gzip only). The header doesn't let you say "please not gzip", so if they mention it, then they support it (even if it is not preferred).

It is true that this is not ideal for future multiple compression types.
Especially if we wish to honor the browser's preference order.

Reply:
Per RFC 7231, section 5.3.4. Accept-Encoding

> 2.  If the representation has no content-coding, then it is
       acceptable by default unless specifically excluded by the
       Accept-Encoding field stating either "identity;q=0" or "*;q=0"
       without a more specific entry for "identity".

>   3.  If the representation's content-coding is one of the
       content-codings listed in the Accept-Encoding field, then it is
       acceptable unless it is accompanied by a qvalue of 0.  (As
       defined in [Section 5.3.1](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.1), a qvalue of 0 means "not acceptable".)
       
 So if it specifies `Accept-Encoding: gzip;q=0` it's *denying* gzip

Reply:
Right you are, I glossed right over that.  I'll see if I can implement something that makes more sense than "contains".

Reply:
Have a look here: https://github.com/rwf2/Rocket/blob/a866134212a13bfaa61a1cd1e68880461af1ab42/core/http/src/parse/media_type.rs. That’s a very similar logic, and this directory should also be the appropriate place to add this parsing code.

Reply:
This was considerably more work than I had hoped.

Reply:
@palant I would gladly take any additional unit tests, my time I can spend on this project is running short.

Reply:
@hcldan Unfortunately, it’s the same situation here. Unless I see some PRs being accepted in a meaningful form, I have no more time left to contribute to this project.